### PR TITLE
auth: add serviceaccount proxy

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This PR will add a proxy for the service account service. 

**Why do we need this feature?**

This proxy will be used to protect service accounts created by an external service account provider.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by gamab — COMMENTED** [review]
Looking good!
Don't forget to protect the tokens as well :)

File: pkg/services/serviceaccounts/proxy/service.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		log:            log.New("serviceaccounts.proxy"),
+		proxiedService: proxiedService,
+	}
+	return s, nil
+}
+
+var _ serviceaccounts.Service = (*ServiceAccountsProxy)(nil)
+
+func (s *ServiceAccountsProxy) CreateServiceAccount(ctx context.Context, orgID int64, saForm *serviceaccounts.CreateServiceAccountForm) (*serviceaccounts.ServiceAccountDTO, error) {
+	if isExternalServiceAccount(saForm.Name) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:36]: Since the check is on the name not the generated login, we shouldn't check against `sa-extsvc-` but against `extsvc-`

File: pkg/services/serviceaccounts/proxy/service.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, err
+	}
+
+	sa.IsExternal = isExternalServiceAccount(sa.Login)
+
+	return sa, nil
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error) {
+	return s.proxiedService.RetrieveServiceAccountIdByName(ctx, orgID, name)
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:67]: Should we set `IsExternal` here as well?

File: pkg/services/serviceaccounts/proxy/service.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func (s *ServiceAccountsProxy) UpdateServiceAccount(ctx context.Context, orgID, serviceAccountID int64, saForm *serviceaccounts.UpdateServiceAccountForm) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	if isExternalServiceAccount(*saForm.Name) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:71]: Same remark the check is on the name not the generated login, we shouldn't check against `sa-extsvc-` but against `extsvc-`

File: pkg/services/serviceaccounts/proxy/service.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, err
+	}
+	if isExternalServiceAccount(sa.Login) {
+		return nil, extsvcaccounts.ErrExtServiceAccountCannotBeUpdated
+	}
+
+	return s.proxiedService.UpdateServiceAccount(ctx, orgID, serviceAccountID, saForm)
+}
+
+func (s *ServiceAccountsProxy) AddServiceAccountToken(ctx context.Context, serviceAccountID int64, cmd *serviceaccounts.AddServiceAccountTokenCommand) (*apikey.APIKey, error) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:85]: We can do that in another PR to keep changes small. But the service accounts API is by-passing the service to delete tokens.
We need a `DeleteServiceAccountToken` function and we need it to go through the proxy.

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+	"testing"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/extsvcauth/extsvcaccounts"
+	"github.com/grafana/grafana/pkg/services/serviceaccounts"
+	"github.com/stretchr/testify/assert"
+)
+
+type FakeServiceAccountsService struct {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:14]: Instead of re-creating this, I'd suggest moving the Fake from: `pkg/services/serviceaccounts/api/api_test.go` to `pkg/services/serviceaccounts/tests` add use it in both API tests and Proxy tests.

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+					Name: "my-service-account",
+				},
+				expectedError: nil,
+			},
+			{
+				description: "should not allow to create a service account with sa-extsvc prefix",
+				form: serviceaccounts.CreateServiceAccountForm{
+					Name: "sa-extsvc-my-service-account",
+				},
+				expectedError: extsvcaccounts.ErrExtServiceAccountInvalidName,
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:77]:
```suggestion
				description: "should not allow to create a service account with `extsvc-` prefix",
				form: serviceaccounts.CreateServiceAccountForm{
					Name: "extsvc-my-service-account",
				},
				expectedError: extsvcaccounts.ErrExtServiceAccountInvalidName,
```

**Review by gamab — COMMENTED** [review]
Great job there!
I left a couple of minor suggestions. 
I guess the most important ones are: 
* service accounts' "name" shouldn't be prefixed by `extsvc-` (not `sa-extsvc`).
* we are missing the remove/delete token function in the service accounts service interface.

File: pkg/services/serviceaccounts/proxy/service.go (hunk 2 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go:67]: Not really, it will be ignored  since the return value is `int64`

File: pkg/services/serviceaccounts/proxy/service_test.go (hunk 5 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service_test.go:14]: This is a great idea, but do you think it is within the scope of this pr? I don't mind do this in a follow-up PR just to keep a clear scoped between PRs, rather than moving the goalpost

File: pkg/services/serviceaccounts/proxy/service.go (hunk 1 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go]: Changed as requested

File: pkg/services/serviceaccounts/proxy/service.go (hunk 3 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go:73]: Changed as requested

File: pkg/services/serviceaccounts/proxy/service.go (hunk 2 above)
Comment (gamab) [reply pkg/services/serviceaccounts/proxy/service.go:69]: My bad, I didn't check the return type :+1:

File: pkg/services/serviceaccounts/proxy/service.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	return sa, nil
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error) {
+	return s.proxiedService.RetrieveServiceAccountIdByName(ctx, orgID, name)
+}
+
+func (s *ServiceAccountsProxy) UpdateServiceAccount(ctx context.Context, orgID, serviceAccountID int64, saForm *serviceaccounts.UpdateServiceAccountForm) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	if isExternalServiceAccount(*saForm.Name) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:73]:
```suggestion
	if isNameValid(*saForm.Name) {
```

File: pkg/services/serviceaccounts/proxy/service.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+		proxiedService: proxiedService,
+	}
+	return s, nil
+}
+
+var _ serviceaccounts.Service = (*ServiceAccountsProxy)(nil)
+
+func (s *ServiceAccountsProxy) CreateServiceAccount(ctx context.Context, orgID int64, saForm *serviceaccounts.CreateServiceAccountForm) (*serviceaccounts.ServiceAccountDTO, error) {
+	if isNameValid(saForm.Name) {
+		s.log.Info("Unable to create service account with a protected name", "name", saForm.Name)
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:37]:
```suggestion
		s.log.Error("Unable to create service account with a protected name", "name", saForm.Name)
```

File: pkg/services/serviceaccounts/proxy/service_test.go (hunk 5 above)
Comment (gamab) [reply pkg/services/serviceaccounts/proxy/service_test.go:14]: Depends how big of a change it is, but if it's only three files, let's do it once and for all now  :100:

File: pkg/services/serviceaccounts/proxy/service.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+		s.log.Info("unable to create tokens for external service accounts", "serviceAccountID", serviceAccountID)
+		return nil, extsvcaccounts.ErrCannotCreateToken
+	}
+
+	return s.proxiedService.AddServiceAccountToken(ctx, serviceAccountID, cmd)
+}
+
+func isNameValid(name string) bool {
+	return strings.HasPrefix(name, extsvcaccounts.ExtsvcPrefix)
+}
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:105]: I think the logic needs to be reversed here
```suggestion
func isNameValid(name string) bool {
	return !strings.HasPrefix(name, extsvcaccounts.ExtsvcPrefix)
}
```

File: pkg/services/serviceaccounts/proxy/service.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		log:            log.New("serviceaccounts.proxy"),
+		proxiedService: proxiedService,
+	}
+	return s, nil
+}
+
+var _ serviceaccounts.Service = (*ServiceAccountsProxy)(nil)
+
+func (s *ServiceAccountsProxy) CreateServiceAccount(ctx context.Context, orgID int64, saForm *serviceaccounts.CreateServiceAccountForm) (*serviceaccounts.ServiceAccountDTO, error) {
+	if isNameValid(saForm.Name) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service.go:36]:
```suggestion
	if !isNameValid(saForm.Name) {
```

File: pkg/services/serviceaccounts/proxy/service.go (hunk 7 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go]: changed

File: pkg/services/serviceaccounts/proxy/service.go (hunk 9 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go]: changed

File: pkg/services/serviceaccounts/proxy/service.go (hunk 10 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go]: changed

File: pkg/services/serviceaccounts/proxy/service.go (hunk 8 above)
Comment (linoman) [reply pkg/services/serviceaccounts/proxy/service.go]: changed

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+	})
+
+	t.Run("should retrieve service account with IsExternal field", func(t *testing.T) {
+		testCases := []struct {
+			description            string
+			expectedServiceAccount *serviceaccounts.ServiceAccountProfileDTO
+			expectedIsExternal     bool
+		}{
+			{
+				description: "should allow to delete a service account",
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:128]:
```suggestion
				description: "should not mark as external",
```

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+				expectedServiceAccount: &serviceaccounts.ServiceAccountProfileDTO{
+					Login: "my-service-account",
+				},
+				expectedIsExternal: false,
+			},
+			{
+				description: "should not allow to delete a service account with sa-extsvc prefix",
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:135]:
```suggestion
				description: "should mark as external",
```

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+				form: serviceaccounts.UpdateServiceAccountForm{
+					Name: &nameWithoutProtectedPrefix,
+				},
+				expectedServiceAccount: &serviceaccounts.ServiceAccountProfileDTO{
+					Login: "my-service-account",
+				},
+				expectedError: nil,
+			},
+			{
+				description: "should not allow to update a non-external service account with sa-extsvc prefix",
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:173]:
```suggestion
				description: "should not allow to update a non-external service account with extsvc prefix",
```

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+				form: serviceaccounts.UpdateServiceAccountForm{
+					Name: &nameWithoutProtectedPrefix,
+				},
+				expectedServiceAccount: &serviceaccounts.ServiceAccountProfileDTO{
+					Login: "sa-extsvc-my-service-account",
+				},
+				expectedError: extsvcaccounts.ErrCannotBeUpdated,
+			},
+			{
+				description: "should not allow to update an external service account with a sa-extsvc prefix",
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:193]:
```suggestion
				description: "should not allow to update an external service account with a extsvc prefix",
```

File: pkg/services/serviceaccounts/proxy/service_test.go [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+			t.Run(tc.description, func(t *testing.T) {
+				tc := tc
+				serviceMock.ExpectedServiceAccountProfileDTO = tc.expectedServiceAccount
+				_, err := svc.AddServiceAccountToken(context.Background(), testServiceAccountId, &tc.cmd)
+				assert.Equal(t, tc.expectedError, err, tc.description)
+			})
+		}
+	})
+
+	t.Run("should identify service account names for being external or not", func(t *testing.T) {
```
Comment (gamab) [inline pkg/services/serviceaccounts/proxy/service_test.go:253]:
```suggestion
	t.Run("should identify service account logins for being external or not", func(t *testing.T) {
```

**Review by gamab — COMMENTED** [review]
Couple of tests comments
Also as we discussed seems like the proxy service is not getting injected in the API as it's not matching the API service interface.
Maybe we should merge this without the binding yet, work a bit on the interface problem and add the binding  back?

**Comment by linoman** [comment]
> Couple of tests comments Also as we discussed seems like the proxy service is not getting injected in the API as it's not matching the API service interface. Maybe we should merge this without the binding yet, work a bit on the interface problem and add the binding back?

Comments were addressed and as we discussed, I removed the wiring. Let's get the API service interface in order and then we can wired this properly.

File: pkg/services/serviceaccounts/models.go [hunk 16]
```diff
@@ -146,7 +150,10 @@ type ServiceAccountProfileDTO struct {
 	// example: Editor
 	Role string `json:"role" xorm:"role"`
 	// example: []
-	Teams         []string        `json:"teams" xorm:"-"`
+	Teams []string `json:"teams" xorm:"-"`
+	// example: false
+	IsExternal bool
```
Comment (gamab) [inline pkg/services/serviceaccounts/models.go:155]:
```suggestion
	IsExternal bool `json:"isExternal,omitempty" xorm:"-"`
```

**Review by gamab — APPROVED** [review]

**Comment by gamab** [comment]
Good job Lino! :tada:
